File: api/src/controllers/chat.controller.ts
import { Request, Response } from 'express';
import { RAGService } from '../services/rag.service';
import { ChatMessage } from '../models/database.model';
import { expandQuery } from "../services/query-expansion.service";

export class ChatController {
  constructor(private ragService: RAGService) { }

  /**
   * Chat dạng realtime (UI chat)
   */
  async chat(req: Request, res: Response): Promise<void> {
    try {
      const { messages } = req.body;

      if (!Array.isArray(messages) || messages.length === 0) {
        res.status(400).json({ botMessage: 'Tin nhắn không hợp lệ' });
        return;
      }

      // Chuẩn hóa message
      const validMessages: ChatMessage[] = messages.map((m: any) => ({
        role: m.role === 'user' ? 'user' : 'assistant',
        content: String(m.content || ''),
      }));

      /**
       * Gọi RAG service
       */
      const botMessage = await this.ragService.chat(validMessages);

      res.json({ botMessage });
    } catch (error) {
      console.error('Chat error:', error);
      res.status(500).json({
        botMessage: 'Xin lỗi, hệ thống đang gặp lỗi.',
      });
    }
  }

  /**
   * Endpoint tích hợp với repo course-ai
   */
  async ask(req: Request, res: Response): Promise<void> {
    try {
      // Support the UI chat payload too: { messages: [{ role, content }, ...] }
      // This effectively merges /api/chat into /api/demo_agent/v1/ask.
      if (Array.isArray(req.body?.messages)) {
        await this.chat(req, res);
        return;
      }

      const { session_id, model_id, user, prompt, context } = req.body;

      if (!session_id || !model_id || !user || !prompt) {
        res.status(400).json({
          status: 'error',
          message: 'Thiếu dữ liệu bắt buộc: session_id, model_id, user, prompt',
        });
        return;
      }

      const history = Array.isArray(context?.history) ? context.history : [];

      const messages: ChatMessage[] = [
        ...history.map((m: any) => ({
          role: m.role === 'user' ? 'user' : 'assistant',
          content: String(m.content || ''),
        })),
        {
          role: 'user',
          content: String(prompt),
        },
      ];

      const startTime = Date.now();
      const content = await this.ragService.chat(messages);
      const responseTimeMs = Date.now() - startTime;

      res.json({
        session_id,
        status: 'success',
        content_markdown: content,
        meta: {
          model: model_id,
          response_time_ms: responseTimeMs,
          tokens_used: 0,
        },
        attachments: [],
      });
    } catch (error) {
      console.error('Ask error:', error);
      res.status(500).json({
        session_id: req.body?.session_id,
        status: 'error',
        message: 'Xin lỗi, hệ thống đang gặp lỗi.',
      });
    }
  }
}

File: api/backend/controllers/chat_controller.py
from __future__ import annotations

import time
from typing import Any

from backend.models.database import ChatMessage
from backend.services.query_expansion_service import expand_query
from backend.services.rag_service import RAGService


class ChatController:
    def __init__(self, rag_service: RAGService) -> None:
        self.rag_service = rag_service

    def chat(self, payload: dict[str, Any]) -> tuple[dict[str, Any], int]:
        messages = payload.get("messages")
        if not isinstance(messages, list) or not messages:
            return {"botMessage": "Invalid message payload"}, 400

        valid_messages: list[ChatMessage] = []
        for item in messages:
            role = "user" if str(item.get("role")) == "user" else "assistant"
            content = str(item.get("content") or "")
            valid_messages.append(ChatMessage(role=role, content=content))

        last_message = valid_messages[-1]
        expanded_queries: list[str] | None = None
        if last_message.role == "user":
            expanded_queries = expand_query(last_message.content)

        try:
            bot_message = self.rag_service.chat(valid_messages, expanded_queries)
            return {"botMessage": bot_message}, 200
        except Exception as exc:
            return {
                "botMessage": "System error while processing your request",
                "error_detail": f"{type(exc).__name__}: {exc}",
            }, 500

    def ask(self, payload: dict[str, Any]) -> tuple[dict[str, Any], int]:
        if isinstance(payload.get("messages"), list):
            return self.chat(payload)

        session_id = payload.get("session_id")
        model_id = payload.get("model_id")
        user = payload.get("user")
        prompt = payload.get("prompt")
        context = payload.get("context") or {}

        if not session_id or not model_id or not user or not prompt:
            return (
                {
                    "status": "error",
                    "message": "Missing required fields: session_id, model_id, user, prompt",
                },
                400,
            )

        history = context.get("history") if isinstance(context, dict) else []
        if not isinstance(history, list):
            history = []

        messages: list[ChatMessage] = []
        for item in history:
            role = "user" if str(item.get("role")) == "user" else "assistant"
            content = str(item.get("content") or "")
            messages.append(ChatMessage(role=role, content=content))

        messages.append(ChatMessage(role="user", content=str(prompt)))
        expanded_queries = expand_query(str(prompt))

        start_time = time.time()
        try:
            content_markdown = self.rag_service.chat(messages, expanded_queries)
            response_time_ms = int((time.time() - start_time) * 1000)
            return (
                {
                    "session_id": session_id,
                    "status": "success",
                    "content_markdown": content_markdown,
                    "meta": {
                        "model": model_id,
                        "response_time_ms": response_time_ms,
                        "tokens_used": 0,
                    },
                    "attachments": [],
                },
                200,
            )
        except Exception as exc:
            return (
                {
                    "session_id": session_id,
                    "status": "error",
                    "message": "System error while processing your request",
                    "error_detail": f"{type(exc).__name__}: {exc}",
                },
                500,
            )
